### DISCUSSION TITLE
Invoke-Deploy - Change Destination

### DISCUSSION THREAD

**Discussion by lcfut** [CrowdStrike/psfalcon#105] [Q&A] [answered]
I am using the Invoke-Deploy command to send my executable to select endpoints.
How can I change the destination location where the executable lands and runs from?

**Comment by bk-cs** [comment, marked as answer]
`Invoke-FalconDeploy` combines the `put` and `run` Real-time Response commands. `put` does not provide the ability to specify a destination--it uploads the file to the current directory. If you want it to be in a different directory, you have to `cd` to that location first.

`Invoke-FalconDeploy` was created as a mechanism to deploy the Falcon Forensics executable, so I don't really have any intention of modifying it, especially because it works for most "upload and run this exe" use cases.

If you'd like to use a more complicated workflow (like using `put` to place the file in a specific directory), I recommend either using a combination of commands with `Invoke-FalconRTR` or creating a script with the necessary commands to step through each of the parts of what you're trying to do. It might look something like...

```
Start-FalconSession ...
Send-FalconCommand -Command cd ...
Send-FalconAdminCommand -Command put ...
Send-FalconAdminCommand -Command run ...
```

Then you'll have to keep track of the responses and update accordingly. You can see an example of how to do this by looking at the code within `Invoke-FalconRTR` (the `Invoke-RTR` function in `Public\Scripts.ps1`) or `Invoke-FalconDeploy` (`Invoke-Deploy` in the same `Scripts.ps1` file) if you'd like to output the results to CSV.

**Comment by lcfut** [comment]
Thanks for the quick response.
I was using RTR commands like the "cd ......" "put foo.bar" "run foo.bar" and noticed the code was becoming multiple lines of code that could be replaced with the single line of Invoke-Deploy.

Could I start the session - send the command to CD into my folder then Invoke-Deploy and have it run from there?

**Comment by bk-cs** [comment]
Yes. `Invoke-FalconRTR` is meant as a way to run a single Real-time Response command. It creates the session (or batch), sends the command and outputs the results. Although it can be used repeatedly (and the API itself will "append" commands to existing sessions), the "proper" way to do it is to run through each individual step.

You can see examples of that entire workflow on the wiki: https://github.com/CrowdStrike/psfalcon/wiki/Real-time-Response#send-real-time-response-commands-to-a-batch-of-hosts

You'll basically follow those steps, but issue multiple commands to a single (or batch) session. If it's something that will take longer than the maximum timeout of 600 seconds (which defaults to 30 seconds if you don't specify it when issuing the command), you'll also need to keep a timer going and `Update-FalconSession` periodically to ensure the session running on each host does not expire.

You'll also want to add logic to only continue if the previous command was successful, so you don't end up doing things like using `put` in the wrong directory. You can see some examples of this with `Invoke-FalconDeploy` where it only selects successful hosts and then specifies the `OptionalHostIds` field on the next command.

**Comment by oracle701** [comment]
Hello,
I have a similar need to have Invoke-falcondeploy PUT the file into a specific location, unless there is another option.
What I need is a command to run an exe with a couple arguments. RUN doesn't seem to have that option and 'runscript' seems to stop the exe after a couple of seconds. 

Is there another way?

**Reply by bk-cs** [reply]
`Invoke-FalconDeploy` is designed to do this, but the destination is a temporary folder which is not modifiable. Will your executable only work if it's in a specific location?

**Reply by oracle701** [reply]
I am trying to run a disk copy task during an incident of the internal drive to a new partition\drive. I am trying to create the folder for that application and run the exe on that drive only, not to the C drive then copy to new partition\drive.
So yes, it will only work from a specific location in the sense that we don't want to modify the C drive before a copy is made.

So far, what I have does make the folder but does not copy the zip file to D:\TEST after, it only copies to C drive.

        $Session = Start-FalconSession -HostId $AID

        $Command = Invoke-FalconAdminCommand -Command mkdir -Arguments D:\TEST -SessionId $Session.session_id
        Confirm-FalconCommand -CloudRequestId $Command.cloud_request_id

        $Command = Invoke-FalconCommand -Command cd -Arguments D:\TEST -SessionId $Session.session_id
        Confirm-FalconCommand -CloudRequestId $Command.cloud_request_id

        $Command = Invoke-FalconAdminCommand -Command put FTKImager.zip -SessionId $Session.session_id
        Confirm-FalconCommand -CloudRequestId $Command.cloud_request_id

**Reply by bk-cs** [reply]
You can't continue with the next step unless the previous command has been successfully completed. You can fix that by doing one of two things...

Add logic to check if the previous command completed successfully before continuing:
```
try {
    $Session = Start-FalconSession -HostId $AID
    if ($Session) {
        $Mkdir = Invoke-FalconAdminCommand -Command mkdir -Argument D:\TEST -SessionId $Session.session_id
        if ($Mkdir.cloud_request_id) {
            do {
                Start-Sleep 5
                $ConfirmMkdir = Confirm-FalconCommand -CloudRequestId $Mkdir.cloud_request_id
            } until ($ConfirmMkdir.complete -eq $true)
            if ($ConfirmMkdir.complete -eq $true) {
                $Cd = Invoke-FalconCommand -Command cd -Argument D:\TEST -SessionId $Session.session_id
                if ($Cd.cloud_request_id) {
                    do {
                        Start-Sleep 5
                        $ConfirmCd = Confirm-FalconCommand -CloudRequestId $Cd.cloud_request_id
                    } until ($ConfirmCd.complete -eq $true)
                    if ($ConfirmCd.complete -eq $true) {
                        $Put = Invoke-FalconAdminCommand -Command put -Argument FTKImager.zip -SessionId $Session.session_id
                        if ($Put.cloud_request_id) {
                            do {
                                Start-Sleep 5
                                $ConfirmPut = Confirm-FalconCommand -CloudRequestId $Put.cloud_request_id
                            } until ($ConfirmPut.complete -eq $true)
                            $ConfirmPut
                        }
                    }
                }
            }    
        }
    }   
} catch {
    throw $_
}
```
Convert to a batch session which does not require confirmation:
```
try {
    $Batch = Start-FalconSession -HostId $AID -Timeout 600
    if ($Batch) {
        $Mkdir = Invoke-FalconAdminCommand -Command mkdir -Argument D:\TEST -BatchId $Batch.batch_id -Timeout 600
        if ($Mkdir.complete -eq $true) {
            $Cd = Invoke-FalconCommand -Command cd -Argument D:\TEST -BatchId $Batch.batch_id -Timeout 600
            if ($Cd.complete -eq $true) {
                Invoke-FalconAdminCommand -Command put -Argument FTKimager.zip -BatchId $Batch.batch_id -Timeout 600   
            }
        }
    }
} catch {
    throw $_
}
```
**EDIT**: Added `do/until` loops.

**Reply by oracle701** [reply]
That works, thanx! 
Now that the exe is where it should be, it runs but stops after about 5-10 seconds. I have tried this with other exe's too, and directly from the RTR console. How can I let it run until it completes?

**Reply by bk-cs** [reply]
You need to run it as a secondary process so that RTR doesn't attempt to wait for it to complete. Try this:

```
Invoke-FalconRtr -Command runscript -Argument '-Raw=```powershell.exe -c "&{ Start-Process -FilePath `"D:\TEST\my.exe`" -ArgumentList `"my command arguments`"}"```
```

Note that the double quotes inside the brackets are escaped with a backtick.

**Reply by oracle701** [reply]
From my testing, that method does make the exe run but is not running any longer then before. It also doesn't allow multiple arguments in -ArgumentList.
For example "MSERT.exe /f /q" (full scan, quiet) or with my original exe, "ftkimager.exe \\.\PhysicalDrive0 D:\Test01 --case-number 01".

**Reply by bk-cs** [reply]
> From my testing, that method does make the exe run but is not running any longer then before. It also doesn't allow multiple arguments in -ArgumentList. For example "MSERT.exe /f /q" (full scan, quiet) or with my original exe, "ftkimager.exe \.\PhysicalDrive0 D:\Test01 --case-number 01".

Can you provide a full syntax example? My guess is that something is broken within the quoting.

**Reply by oracle701** [reply]
If /f is removed from the last line, it runs but still stops after after few seconds. With 2 or more arguments, it fails.      


```
$Batch = Start-FalconSession -HostId $AID -Timeout 600
        if ($Batch) {
            $Mkdir = Invoke-FalconAdminCommand -Command mkdir -Argument D:\TEST -BatchId $Batch.batch_id -Timeout 600
            if ($Mkdir.complete -eq $True) {
                $Cd = Invoke-FalconCommand -Command cd -Argument D:\TEST -BatchId $Batch.batch_id -Timeout 600
                if ($Cd.complete -eq $True) {
                    $Put = Invoke-FalconAdminCommand -Command put MSERT.zip -BatchId $Batch.batch_id -Timeout 600
                    if ($Put.complete -eq $True) {
                        $Unzip = Invoke-FalconAdminCommand -Command runscript -Arguments '-Raw=```Expand-Archive MSERT.zip -DestinationPath D:\TEST\```' -BatchId $Batch.batch_id -Timeout 600
                        if ($Unzip.complete -eq $True) {
                            $Cd2 = Invoke-FalconCommand -Command cd -Arguments D:\TEST -BatchId $Batch.batch_id -Timeout 600
                            if ($Cd2.complete -eq $True) {
                                Invoke-FalconRtr -Command runscript -Argument '-Raw=```powershell.exe -c "&{ Start-Process -FilePath `"D:\TEST\msert.exe`" -ArgumentList `"/f /q`"}"```' -HostId $AID -Timeout 600
                            }
                        }
                    }
                }
            }
        
        
        }
```

**Reply by bk-cs** [reply]
Does the executable behave differently if you run the same command locally within PowerShell?

```
powershell.exe -c "&{ Start-Process -FilePath `"D:\TEST\msert.exe`" -ArgumentList `"/f /q`" }"
```

**Reply by oracle701** [reply]
Thank you for that question, led me to the answer.
Yes, I did receive the same error with that command and exe did not run.

Did a bit of research and this is the proper syntax for multiple arguments and letting the exe run.
I need `" before and after each argument, and separated by a comma.

For example:
Invoke-FalconRtr -Command runscript -Argument '-Raw=```powershell.exe -c "&{ Start-Process -FilePath `"D:\msert.exe`" -ArgumentList `"/f`",`"/q`"}"```' -HostId $AID -Timeout 600

So for anyone else following along, this is the final code that is working for me. 

        $Batch = Start-FalconSession -HostId $AID -Timeout 600
        if ($Batch) {
            $Mkdir = Invoke-FalconAdminCommand -Command mkdir -Argument D:\TEST -BatchId $Batch.batch_id -Timeout 600
            if ($Mkdir.complete -eq $True) {
                $Cd = Invoke-FalconCommand -Command cd -Argument D:\TEST -BatchId $Batch.batch_id -Timeout 600
                if ($Cd.complete -eq $True) {
                    $Put = Invoke-FalconAdminCommand -Command put MSERT.zip -BatchId $Batch.batch_id -Timeout 600
                    if ($Put.complete -eq $True) {
                        $Unzip = Invoke-FalconAdminCommand -Command runscript -Arguments '-Raw=```Expand-Archive MSERT.zip -DestinationPath D:\TEST\```' -BatchId $Batch.batch_id -Timeout 600
                        if ($Unzip.complete -eq $True) {
                            $Cd2 = Invoke-FalconCommand -Command cd -Arguments D:\TEST -BatchId $Batch.batch_id -Timeout 600
                            if ($Cd2.complete -eq $True) {
                                Invoke-FalconRtr -Command runscript -Argument '-Raw=```powershell.exe -c "&{ Start-Process -FilePath `"D:\TEST\msert.exe`" -ArgumentList `"/f`",`"/q`"}"```' -HostId $AID -Timeout 600
                            }
                        }
                    }
                }
            }
        
        
        }

**Reply by oracle701** [reply]
Hey,
I was able to figure out the syntax for parameters with a space (`\""--compress 5\`"), but what is the syntax for putting in a variable instead?

Example: Invoke-FalconRTR -Command runscript -Argument '-Raw=```powershell.exe -c "&{ Start-Process -FilePath `"D:\TEST\ftkimager\ftkimager.exe`" -ArgumentList `"$DRIVE`",`"D:\Test01`",`"--e01`",`\""--compress 5\`"}"```' -HostId $AID

**Reply by bk-cs** [reply]
Try this:

```powershell
$Raw = "powershell.exe -c `"&{ Start-Process -FilePath 'D:\TEST\ftkimager\ftkimager.exe' -ArgumentList '$DRIVE','D:\Test01','--e01','--compress 5'}`""
$Raw = '```' + $Raw + '```'
Invoke-FalconRTR -Command runscript -Argument $Raw -HostId $AID
```

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules?view=powershell-7.2

**Reply by oracle701** [reply]
With that I get an error. Have not found a solution yet.
"40012: Max args is 0. 1 were provided"

**Reply by bk-cs** [reply]
Does this work?
```powershell
$DRIVE = 'C:\my\folder'
$Raw = 'powershell.exe -c "&{ Start-Process -FilePath `"D:\TEST\ftkimager\ftkimager.exe`" -ArgumentList `"{0}`",`"D:\Test01`",`"--e01`",`"--compress 5`"}"' -f $DRIVE
$Raw = '```' + $Raw + '```'
Invoke-FalconRTR -Command runscript -Argument $Raw -HostId $AID
```

You could also create a script and provide `$DRIVE` when launching it:

example.ps1
```powershell
param([Parameter(Mandatory=$true,Position=1)][string]$Variable)
powershell.exe -c "&{ Start-Process -FilePath 'D:\TEST\ftkimager\ftkimager.exe' -ArgumentList '$Variable','D:\Test01','--e01','--compress 5' }"
```
```powershell
$DRIVE = 'C:\my\folder'
Invoke-FalconRtr runscript '-CloudFile="example.ps1"' "-CommandLine='$DRIVE'" -HostId $AID
```

**Reply by oracle701** [reply]
First one does not. I am still working on the script test for the 2nd one though, busy week.

**Reply by oracle701** [reply]
Success!
I kept playing around with your example of:

```
$Raw = "powershell.exe -c `"&{ Start-Process -FilePath 'D:\TEST\ftkimager\ftkimager.exe' -ArgumentList '$DRIVE','D:\Test01','--e01','--compress 5'}`""
$Raw = '```' + $Raw + '```'
Invoke-FalconRTR -Command runscript -Argument $Raw -HostId $AID
```

I finally go it to work with a variable by adding -Raw= to the beginning of it. So now I have:

```
$Raw = "powershell.exe -c `"&{ Start-Process -FilePath 'D:\TEST\ftkimager\ftkimager.exe' -ArgumentList '$DRIVE','D:\Test01','--e01','--compress 5'}`""
$Raw = '-Raw=' + '```' + $Raw + '```'
Invoke-FalconRTR -Command runscript -Argument $Raw -HostId $AID -Timeout 600
```

Thanx!

**Reply by bk-cs** [reply]
Sorry, that was a total typo on my part! Nice job.
